torch/_lobpcg.py

```diff
@@ -0,0 +1,719 @@
+            dd = d d^T                         # this is 2-d matrix
+            R^T R = Cholesky(dd * SBS)
+            Ri = R^-1 * d                      # broadcasting
+            solve symeig problem Ri^T SAS Ri Z = Theta Z
+            C = Ri Z
+
+          where `dd` is 2-d matrix that replaces matrix products `D M
+          D` with one element-wise product `M * dd`; and `d` replaces
+          matrix product `D M` with element-wise product `M *
+          d`. Also, creating the diagonal matrix `D` is avoided.
+
+        Arguments:
+        S (Tensor): the matrix basis for the search subspace, size is
+                    :math:`(m, n)`.
+
+        Returns:
+        Ri (tensor): upper-triangular transformation matrix of size
+                     :math:`(n, n)`.
+
+        """
+        B = self.B
+        mm = torch.matmul
+        SBS = _utils.qform(B, S)
+
+        d1 = SBS.diagonal(0, -2, -1) ** -0.5
+        if d1.is_cuda:
+            # Avoid `test_torch.TestTorchDeviceTypeCUDA.test_lobpcg_basic_cuda_float64 leaked 4096 bytes CUDA memory on device 0`
+            d = torch.zeros((d1.shape[0], 1), dtype=d1.dtype, device=d1.device)
+            d[:, 0] = d1
+        else:
+            d = d1.reshape(d1.shape[0], 1)
+
+        dd = mm(d, _utils.transpose(d))
+        R = torch.cholesky(dd * SBS, upper=True)
+        # TODO: could use LAPACK ?trtri as R is upper-triangular
+        Rinv = torch.inverse(R)
+        return Rinv * d
+
+    def _get_svqb(self,
+                  U,     # Tensor
+                  drop,  # bool
+                  tau    # float
+                  ):
+        """Return B-orthonormal U.
+
+        .. note:: When `drop` is `False` then `svqb` is based on the
+                  Algorithm 4 from [DuerschPhD2015] that is a slight
+                  modification of the corresponding algorithm
+                  introduced in [StathopolousWu2002].
+
+        Arguments:
+
+          U (Tensor) : initial approximation, size is (m, n)
+          drop (bool) : when True, drop columns that
+                     contribution to the `span([U])` is small.
+          tau (float) : positive tolerance
+
+        Returns:
+
+          U (Tensor) : B-orthonormal columns (:math:`U^T B U = I`), size
+                       is (m, n1), where `n1 = n` if `drop` is `False,
+                       otherwise `n1 <= n`.
+
+        """
+        # type: (Tensor, bool, float) -> Tensor
+        if torch.numel(U) == 0:
+            return U
+        UBU = _utils.qform(self.B, U)
+        d = UBU.diagonal(0, -2, -1)
+        # detect and drop zero columns from U
+        nz = torch.where(abs(d) != 0.0)
+        assert len(nz) == 1, nz
+        if len(nz[0]) < len(d):
+            U = U[:, nz[0]]
+            if torch.numel(U) == 0:
+                return U
+            UBU = _utils.qform(self.B, U)
+            d = UBU.diagonal(0, -2, -1)
+            nz = torch.where(abs(d) != 0.0)
+            assert len(nz[0]) == len(d)
```

Comment:
No need to recompute `UBU` after the removal. Sufficient to remove the `i`-th columns and the `i`-th row from the full `UBU` from above.

Reply:
I agree. However, taking into account that (i) the above condition is rarely satisfied and (ii) the removal of possibly several columns and rows from a matrix is not simpler to express as just for recomputing `UBU`, I would leave it as it is for clarity. What do you think?

Reply:
But it looks like you eliminate the columns corresponding to the smallest eigenvalues in the already sorted order? Then it is equivalent to just taking an upper-part of a matrix like UBU(1:nz, 1:nz).

Reply:
No, `d` is not sorted (proof: print out `d` when running the lobpcg unittests).

Reply:
OK, then it is up to you. But it boils down to how you handle small eigenvalues. Me, personally, I'd rather set a tolerance and filter based on that. This way I know what to expect when division by zero occurs.

Reply:
Removing zero columns of `U` has nothing to do with eigenvalues but just correcting the initial approximation of `U`. It's ok to have small diagonal elements in `UBU` as long as these are not exactly zeros that would break the original algorithm.

The original algorithm handles the small eigenvalues, see the following code to the `if`-block.

Reply:
It has, because diagonal elements of UBU are exactly the norms of U in the B-induced metric.

Reply:
the norms squared

Reply:
Agree :)